cdc/redo/manager.go

```diff
@@ -397,75 +358,127 @@ func (m *ManagerImpl) RemoveTable(tableID model.TableID) {
 
 // Cleanup removes all redo logs of this manager, it is called when changefeed is removed
 func (m *ManagerImpl) Cleanup(ctx context.Context) error {
+	m.logBuffer.Close()
+	// We must finish consuming the data here,
+	// otherwise it will cause the channel to not close properly.
+	for range m.logBuffer.Out() {
+		// Do nothing. We do not care about the data.
+	}
+	common.RedoWriteLogDurationHistogram.
+		DeleteLabelValues(m.changeFeedID.Namespace, m.changeFeedID.ID)
+	common.RedoFlushLogDurationHistogram.
+		DeleteLabelValues(m.changeFeedID.Namespace, m.changeFeedID.ID)
 	return m.writer.DeleteAllLogs(ctx)
 }
 
-// updateTableResolvedTs reads rtsMap from redo log writer and calculate the minimum
-// resolved ts of all maintaining tables.
-func (m *ManagerImpl) updateTableResolvedTs(ctx context.Context) error {
-	m.rtsMapMu.Lock()
-	defer m.rtsMapMu.Unlock()
-	rtsMap, err := m.writer.GetCurrentResolvedTs(ctx, m.tableIDs)
-	if err != nil {
-		return err
+func (m *ManagerImpl) flushLog(
+	ctx context.Context,
+	tableRtsMap map[model.TableID]model.Ts,
+	handleErr func(err error),
+) map[model.TableID]model.Ts {
+	if !atomic.CompareAndSwapInt64(&m.flushing, 0, 1) {
+		log.Debug("Fail to update flush flag, " +
+			"the previous flush operation hasn't finished yet")
+		return tableRtsMap
 	}
-	minResolvedTs := uint64(math.MaxUint64)
-	for tableID := range m.rtsMap {
-		if rts, ok := rtsMap[tableID]; ok {
-			m.rtsMap[tableID] = rts
+
+	go func() {
+		defer atomic.StoreInt64(&m.flushing, 0)
+
+		start := time.Now()
+		err := m.writer.FlushLog(ctx, tableRtsMap)
+		if err != nil {
+			handleErr(err)
+			return
```

Comment:
Could it happen this goroutine blocks at somewhere and doesn't exit.

Reply:
It depends on whether the underlying implementation sets timeout. Perhaps we should add some logging to record this exception